### PR TITLE
feat(subscription): add configurable heartbeat for websocket protocol

### LINKED ISSUE #4621: Configure a subgraph websocket ping interval
**Is your feature request related to a problem? Please describe.**
The [graphql-ws](https://github.com/enisdenjo/graphql-ws) protocol supports sending a [`Ping` message](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#ping). This is useful for detecting failed connections, displaying latency metrics or other types of network probing.

If I have opened a passthrough mode subscription to my subgraph the connection is all good. But we're facing issues with longer intervals between messages. Apollo Router logs a TCP connection reset when no messages have been sent beyond our load balancer after a connection idle timeout.

By capturing traffic, we confirmed neither Apollo Router nor the subgraph actively sends pings/heartbeats on the WebSocket session. Both _can_ respond to heartbeats, but both services are just sitting passive and our LB eventually closes the idle connection.

**Describe the solution you'd like**
Add an option to enable sending a `Ping` message to a subgraph from Router on a set interval. Having this off by default is ok, but we need an option to turn it on for our subgraphs

```yaml
subscription:
  enabled: true
  mode:
    passthrough:
      all: 
        path: /subscriptions
        protocol: graphql_ws
        ping_interval: 300 # Some option probably in seconds is fine
```        

**Describe alternatives you've considered**
To keep the connection open we can go and update our subgraph code to also send `Ping` message because this is bi-directional but we may have to do that across many subgraphs and unfortunately some libraries like DGS don't have support to configure a ping interval either

**Additional context**
This is not a concern if we use the [HTTP callback protocol](https://www.apollographql.com/docs/router/executing-operations/subscription-callback-protocol) instead

### PR DESCRIPTION
**Pull request by IvanGoncharov**
Add the ability to enable heartbeat for cases where the subgraph drops idle connections.
For example, https://netflix.github.io/dgs/

Example of configuration:

```yaml
subscription:
  mode:
    passthrough:
      all:
        path: /graphql
        heartbeat_interval: enable #Optional
 ```

Fixes #4621

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Comment by IvanGoncharov** [comment]
Missing unit tests 😞 
I made a few attempts to add them, but each time they ended up being flaky and unreliable.

File: apollo-router/src/protocols/websocket.rs
Comment (IvanGoncharov) [inline apollo-router/src/protocols/websocket.rs:449]: I like my original version with pattern matching: https://github.com/apollographql/router/pull/4802/commits/efcb2bfb10d242ab34cca5598e03df73eeb2d6b8#diff-b932e2ce79bcb30a9685d96419b458d13772deb85c850fe19741532cc21f0dbeL448

But Clippy forced me to rewrite it in this form.

Comment (bnjjj) [reply apollo-router/src/protocols/websocket.rs:449]: I was a little bit confused by the name of the variable. I think you should name it `sink`

File: apollo-router/src/protocols/websocket.rs
Comment (bnjjj) [inline apollo-router/src/protocols/websocket.rs:454]: Should we provide a payload ? If we can avoid sending more data than necessary I would vote for it

File: docs/source/executing-operations/subscription-support.mdx
Comment (bnjjj) [inline docs/source/executing-operations/subscription-support.mdx:104]: Could you add document to explain it also support human readable duration like `10secs`

File: apollo-router/src/protocols/websocket.rs
Comment (IvanGoncharov) [reply apollo-router/src/protocols/websocket.rs:449]: you right! will change

File: apollo-router/src/protocols/websocket.rs
Comment (IvanGoncharov) [reply apollo-router/src/protocols/websocket.rs:454]: It makes debugging easier, without it. 
It's not clear that this is a heartbeat.
Then I looked in traffic I also see RFC6455 PING/PONG so payload makes this particular message to standout.
I think it would also be helpful for customer's engeeneer trying to look into WS traffic.
With default interval of 5s it doesn't add that much overhead.

But this is not a principal issue for me so can remove it if you still have concerns?

File: docs/source/executing-operations/subscription-support.mdx
Comment (IvanGoncharov) [reply docs/source/executing-operations/subscription-support.mdx:104]: will do 👍

File: apollo-router/src/protocols/websocket.rs
Comment (bnjjj) [reply apollo-router/src/protocols/websocket.rs:454]: No I think you're right, if it improves the experience let's keep it

File: apollo-router/src/protocols/websocket.rs
Comment (IvanGoncharov) [reply apollo-router/src/protocols/websocket.rs]: Rewrote this part completely.

File: docs/source/executing-operations/subscription-support.mdx
Comment (IvanGoncharov) [reply docs/source/executing-operations/subscription-support.mdx]: fixed

File: apollo-router/src/plugins/subscription.rs
Comment (bnjjj) [inline apollo-router/src/plugins/subscription.rs:179]: You have to add documentation here

File: apollo-router/src/plugins/subscription.rs
Comment (bnjjj) [inline apollo-router/src/plugins/subscription.rs:174]: Add documentation

File: apollo-router/src/protocols/websocket.rs
Comment (Geal) [inline apollo-router/src/protocols/websocket.rs:464]: technically you don't even need to send with `close_signal`. If it is dropped, `close_sentinel` will resolve and return the error `TryRecvError::Closed`: https://docs.rs/tokio/latest/tokio/sync/oneshot/error/enum.TryRecvError.html
but it is good to be explicit

File: apollo-router/src/protocols/websocket.rs
Comment (Geal) [inline apollo-router/src/protocols/websocket.rs:424]: this is fine to leave as is for now, but as a follow up, we should actually check if we got a response to the heartbeat. Right now it makes sure that the websocket connection stays open, but it does not account for the possibility of the backend hanging

**Review by Geal — APPROVED** [review]
this sounds very reasonable, nicely done. I'd like to get a second review on this before merging, to make sure I did not miss something (I did not sleep much last night)

**Comment by BrynCooke** [comment]
@IvanGoncharov Is there a test that exercises the new behaviour?

File: .changesets/feat_subscription_websocket_heartbeat.md
Comment (abernix) [inline .changesets/feat_subscription_websocket_heartbeat.md:6]:
```suggestion
### Subscriptions: Add configurable "heartbeat" to subgraph WebSocket protocol ([Issue #4621](https://github.com/apollographql/router/issues/4621))

To account for GraphQL Subscription WebSocket implementations (e.g., [DGS](https://netflix.github.io/dgs/)) which drop idle connections by design, the router adds the ability to configure a heartbeat to keep active connections alive.

An example of configuration:
```

File: docs/source/executing-operations/subscription-support.mdx
Comment (abernix) [inline docs/source/executing-operations/subscription-support.mdx:104]: This line is a bit confusing to me.   Could you help me understand it better so I can suggest an alternative?

File: .changesets/feat_subscription_websocket_heartbeat.md
Comment (BrynCooke) [inline .changesets/feat_subscription_websocket_heartbeat.md:13]: Why are we introducing `enable`?
The default before was 5s.

Comment (abernix) [reply .changesets/feat_subscription_websocket_heartbeat.md:13]: I was under the impression that it was disabled by default before?

File: docs/source/executing-operations/subscription-support.mdx
Comment (IvanGoncharov) [reply docs/source/executing-operations/subscription-support.mdx:104]: Ok, so it's disabled by default. 
If you specify human readable time it will be that time, e.g. `10s`
If you specify `enable` it will be `5s`.

File: .changesets/feat_subscription_websocket_heartbeat.md
Comment (IvanGoncharov) [reply .changesets/feat_subscription_websocket_heartbeat.md:13]: Yes, it's disabled by default.
Benjamin suggested to add `enable` since we already have 5s interval as default for callback protocol.
The idea was to allow customers to set reasonable defaults without trying to guess the numbers.

Comment (IvanGoncharov) [reply .changesets/feat_subscription_websocket_heartbeat.md:13]: Enable is equal to 5s, which is our current default for callback protocol.

File: apollo-router/src/protocols/websocket.rs
Comment (IvanGoncharov) [inline apollo-router/src/protocols/websocket.rs:767]: > @IvanGoncharov Is there a test that exercises the new behaviour? As in it should check the heartbeat from the subgraph perspective.

Yes, it checked here. 
1. we stop timer
1. We check that there are no messages in a socket
1. advance timer for duration (60s)
1. check that we got ping
1. check that there are no messages in a socket
1. resume timer

It's done using a saved timer, so Tokio automatically advances the timer, and we don't need to actually wait for the 60s, so the test is fine.
I chose the 60s, so if the mock timer works for some reason, the test will timeout.

**Review by o0Ignition0o — APPROVED** [review]

File: apollo-router/src/plugins/subscription.rs
Comment (IvanGoncharov) [reply apollo-router/src/plugins/subscription.rs:181]: added

File: apollo-router/src/plugins/subscription.rs
Comment (IvanGoncharov) [reply apollo-router/src/plugins/subscription.rs:175]: added

File: docs/source/executing-operations/subscription-support.mdx
Comment (IvanGoncharov) [reply docs/source/executing-operations/subscription-support.mdx]: I rewrote entire comment, hope it's more clear now.

**Pull request merged**
